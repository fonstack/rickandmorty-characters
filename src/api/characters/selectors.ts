import { InfiniteData } from './../../../node_modules/react-query/types/core/types.d';
import { Character, Characters, ServerCharacters, ServerCharacter } from './types';

/**
 * Transforms a single server character to a app character
 */
export const characterTransform = (data: ServerCharacter.ServerResponse): Character => {
  return {
    id: Number(data.id),
    name: data.name,
    image: data.image,
    species: data.species,
    status: data.status.toLowerCase() as Character['status'],
    gender: data.gender.toLowerCase() as Character['gender'],
  };
};

/**
 * Transforms an array of server character to an array of app characters
 */
export const charactersTransform = (data: ServerCharacters.ServerResponse): Characters => {
  return {
    characters: data.characters.map(characterTransform),
    pageInfo: {
      count: data.pageInfo.count,
      next: data.pageInfo.next,
      prev: data.pageInfo.prev,
    },
  };
};

/**
 * Transforms an array (pages) of Characters to app pages of Characters
 */
export const charactersPagesTransform = (data: InfiniteData<ServerCharacters.ServerResponse>): InfiniteData<Characters> => {
  const pages: Characters[] = data.pages.map(charactersTransform);

  return {
    pages,
    pageParams: data.pageParams,
  };
};
