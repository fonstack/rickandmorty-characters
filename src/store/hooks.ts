import { useRecoilState } from 'recoil';
import { favCharactersState } from './atoms';

type useFavUsersStateReturn = [number[], (characterId: number) => void, (characterId: number) => void];

export const useFavUsersState = (): useFavUsersStateReturn => {
  const [favCharacters, setFavCharactersIds] = useRecoilState(favCharactersState);

  const addFavCharacter = (characterId: number) => {
    setFavCharactersIds((prev) => [...prev, characterId]);
  };

  const removeFavCharacter = (characterId: number) => {
    setFavCharactersIds((prev) => prev.filter((id) => id !== characterId));
  };

  return [favCharacters, addFavCharacter, removeFavCharacter];
};
