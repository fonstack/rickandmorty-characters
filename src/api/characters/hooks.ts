import { useInfiniteQuery, UseInfiniteQueryResult, useQuery, UseQueryResult } from 'react-query';
import { getCharactersByIds, getCharactersPaginated } from './api';
import { charactersPagesTransform, characterTransform } from './selectors';
import { Character, Characters, QueryTypes } from './types';

const MAX_PAGES_TO_FETCH = 10;

export const useInfiniteCharacters = (): UseInfiniteQueryResult<Characters> => {
  return useInfiniteQuery(QueryTypes.CHARACTERS, ({ pageParam = 1 }) => getCharactersPaginated(pageParam), {
    getNextPageParam: (lastPage) => {
      if (lastPage.pageInfo.next && lastPage.pageInfo.next < MAX_PAGES_TO_FETCH) {
        return lastPage.pageInfo.next;
      }
      return null;
    },
    select: charactersPagesTransform,
  });
};

export const useCharactersByIds = (ids: number[]): UseQueryResult<Character[]> => {
  return useQuery([QueryTypes.CHARACTERS_BY_IDS, ids.join(',')], () => getCharactersByIds(ids), {
    select: (data) => data.map(characterTransform),
    keepPreviousData: true,
  });
};
