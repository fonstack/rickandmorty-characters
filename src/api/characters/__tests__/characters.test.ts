import { renderHook } from '@testing-library/react-hooks';
import { TestQueryClientProvider } from '../../../tests';
import { useInfiniteCharacters } from '../hooks';
import { charactersTransform, characterTransform } from '../selectors';
import {
  createFakeCharacter,
  createFakeCharacterResponse,
  createFakeCharacters,
  createFakeCharactersResponse,
} from './../mockData';

describe('Characters api tests', () => {
  const wrapper = TestQueryClientProvider;

  it('Character select should transform data properly', () => {
    const fakeCharacterResponse = createFakeCharacterResponse();
    const fakeCharacter = createFakeCharacter();

    expect(characterTransform(fakeCharacterResponse)).toEqual(fakeCharacter);
  });

  it('Characters select should transform data properly', () => {
    const fakeCharactersResponse = createFakeCharactersResponse();
    const fakeCharacters = createFakeCharacters();

    expect(charactersTransform(fakeCharactersResponse)).toEqual(fakeCharacters);
  });

  it('Should get characters correctly', async () => {
    const { result, waitFor, rerender } = renderHook(() => useInfiniteCharacters(), { wrapper });

    await waitFor(() => !result.current.isFetching);
    rerender();

    expect(result.current.data).toMatchSnapshot();
  });
});
