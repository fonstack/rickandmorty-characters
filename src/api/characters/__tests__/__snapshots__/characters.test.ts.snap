// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters api tests Should get characters correctly 1`] = `
Object {
  "pageParams": Array [
    undefined,
  ],
  "pages": Array [
    Object {
      "characters": Array [
        Object {
          "gender": "male",
          "id": 1,
          "image": "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
          "name": "Rick Sanchez",
          "species": "Human",
          "status": "alive",
        },
        Object {
          "gender": "male",
          "id": 2,
          "image": "https://rickandmortyapi.com/api/character/avatar/2.jpeg",
          "name": "Morty Smith",
          "species": "Human",
          "status": "alive",
        },
      ],
      "pageInfo": Object {
        "count": 2,
        "next": 2,
        "prev": null,
      },
    },
  ],
}
`;
